### DISCUSSION TITLE
Road to v1

### DISCUSSION THREAD

**Discussion by bluwy** [bluwy/publint#99] [Announcements]
Making `publint` v1 has been long overdue, mostly because the current abstraction isn't what I think considered "stable". Rules are intentionally written inline and clumsily, as it was easier to prototype and think less about internal structure organization.

For v1, I’d like to focus on extensibility, correctness, and better code structure. This will include:

- Rewrite to TypeScript
- ~~Use changesets~~ (done)
- New context option (https://github.com/bluwy/publint/issues/19)
  - Better environment detection before falling back to context, e.g. analyzing export condition names and `"engines"` field.
- ~~New pack option (https://github.com/bluwy/publint/issues/11)~~ (done)
- New plugin API (https://github.com/bluwy/publint/discussions/101)
  - Rewrite internal rules as individual plugins
- Better programmatic API

I'd also like to move this repo under an org (done) to build up a community around it.
- https://github.com/publint (I'll move the repo here after the initial v1 work is merged into `main`)
- https://www.npmjs.com/org/publint (For future packages)

Additionally, I'd like to improve the site experience (https://github.com/bluwy/publint/discussions/100), but that may take even longer and is not part of the v1 timeline.

While I also foresee this increasing the package size sizably, as long as the total dependency tree doesn't exceed like 1MB, I think it's still acceptable.

## Timeline

~~- 19 July 2024: v1 beta~~
~~- 2 August 2024: v1 stable~~

~~- By end of 2024: Setup changesets and move to publint github org~~
~~- Start of 2024: Tackle #11 (new breaking minor likely required)~~

~~(See update: https://github.com/bluwy/publint/discussions/99#discussioncomment-11633488)~~

## Help

At the moment, I'd like to explore the initial set of features and re-organization myself. Once a beta is ready, all feedback are welcomed. Currently I've done an initial prototype at https://github.com/bluwy/publint/tree/v1-wip

With the new plugin API, I may also need help building some core plugins that handle more cases from feature requests. I'd also want to make sure https://github.com/kravetsone/vscode-publint still works fine.

**Comment by Greenheart** [comment]
Thanks for working on this! It's provided good feedback to help me improve some of my packages ☺️

**Comment by bluwy** [comment]
I've obviously been slacking off but I think it's because I've been trying to take on too much at once. In the next few days, I'll tackle this partly first by migrating to changesets and moving the repo under the publint github org. And then I'll tackle https://github.com/bluwy/publint/issues/11 individually which requires some attention in the short term. I think all these will coincide nicely with 1k stars soon 🙂

**Reply by bluwy** [reply]
This is now done! Updated the discussion description.

**Comment by silvenon** [comment]
I like publint and wanted to ask how it compares to [Are the types wrong?](https://arethetypeswrong.github.io/)

Is it complementary or does it already include the same checks or intends to?

**Reply by bluwy** [reply]
Publint is focused on packaging best practices, so you'd get nudges around "not exporting tests", "proper exports format", etc. Attw is based on TypeScript and reports issues around what TypeScript sees with the package. So the two have slightly different focuses that brings different sets of errors, but it also can happen to overlap, e.g. linting file format types (esm/cjs).

At the moment, it's complementary and I'd recommend running both, but the plan for v1 with a plugin API is also to allow attw to be implemented as a plugin if needed, as what it reports also count as packaging best practices. So maybe we can see both tools converge soon!

**Reply by silvenon** [reply]
Great to hear! I'll be running both then. You're doing great work ❤️
